### PR TITLE
Register CTools Plugins automatically

### PR DESCRIPTION
**Pull request by drupol**
First try to register CTools plugins automatically.

### CODE REVIEW

File: src/ServiceContainer/ServiceProvider/ServiceContainerServiceProvider.php [hunk 1]
```diff
@@ -150,7 +150,34 @@ public function getContainerDefinition() {
       'arguments' => array('cron'),
     );
 
-    // @todo Make it  possible to register all ctools plugins here.
+    if (module_exists('ctools')) {
```
Comment (fabianx-ai) [inline src/ServiceContainer/ServiceProvider/ServiceContainerServiceProvider.php]: should be

$this->moduleExists('ctools')

can't unfortunately use the moduleHandler here not yet.

**Comment by fabianx-ai** [comment]
Otherwise looks great!

File: src/ServiceContainer/ServiceProvider/ServiceContainerServiceProvider.php (hunk 1 above)
Comment (drupol) [inline src/ServiceContainer/ServiceProvider/ServiceContainerServiceProvider.php]: Yes unfortunately....

Maybe if we create a new ServiceContainer plugin especially for CTools ?

File: src/ServiceContainer/ServiceProvider/ServiceContainerServiceProvider.php [hunk 2]
```diff
@@ -150,7 +150,34 @@ public function getContainerDefinition() {
...
+      ctools_include('plugins');
+      $info = ctools_plugin_get_plugin_type_info();
```
Comment (fabianx-ai) [inline src/ServiceContainer/ServiceProvider/ServiceContainerServiceProvider.php]: Same here, should not use procedural code directly, so need a wrapper.

File: src/ServiceContainer/ServiceProvider/ServiceContainerServiceProvider.php [hunk 3]
```diff
@@ -150,7 +150,34 @@ public function getContainerDefinition() {
...
+          ),
+          'calls' => array(
+            array(
+              'setContainer',
+              array(
+                '@service_container',
+              ),
+            ),
+          ),
+        );
```
Comment (fabianx-ai) [inline src/ServiceContainer/ServiceProvider/ServiceContainerServiceProvider.php]: All those lines (the full foreach) should not be necessary, all we need is what is below of here.

File: src/ServiceContainer/ServiceProvider/ServiceContainerServiceProvider.php [hunk 4]
```diff
@@ -150,7 +150,34 @@ public function getContainerDefinition() {
...
+
+        foreach($plugins as $plugin_type => $plugin_data) {
+          $services[$module_name . '.' . $plugin_type] = array();
```
Comment (fabianx-ai) [inline src/ServiceContainer/ServiceProvider/ServiceContainerServiceProvider.php:175]: We should lowercase plugin type for the service and module definition. Or rather:

un-camelCase it

e.g

RenderController => render_controller

There is a handy function to camelCase and convert back in Symfony for that.

File: src/ServiceContainer/ServiceProvider/ServiceContainerServiceProvider.php (hunk 2 above)
Comment (drupol) [inline src/ServiceContainer/ServiceProvider/ServiceContainerServiceProvider.php]: What do you mean ?

File: src/ServiceContainer/ServiceProvider/ServiceContainerServiceProvider.php [hunk 5]
```diff
@@ -150,7 +155,34 @@ public function getContainerDefinition() {
...
+            ),
+          ),
+        );
+
+        foreach($plugins as $plugin_type => $plugin_data) {
+          $plugin_type_clean = $this->toUnderscoreCase($plugin_type);
+          $services[$module_name . '.' . $plugin_type_clean] = array();
+          $parameters['service_container.plugin_managers']['ctools'][$module_name . '.' . $plugin_type_clean] = array(
+            'owner' => $module_name,
+            'type' => $plugin_type_clean,
```
Comment (fabianx-ai) [inline src/ServiceContainer/ServiceProvider/ServiceContainerServiceProvider.php]: This needs to be $plugin_type, not plugin_type_clean.

File: tests/modules/service_container_test_ctools/service_container_test_ctools.info [hunk 6]
```diff
@@ -0,0 +1,9 @@
+name = "Service container test ctools"
+core = 7.x
+
+dependencies[] = service_container
+dependencies[] = registry_autoload
+dependencies[] = ctools
```
Comment (fabianx-ai) [inline tests/modules/service_container_test_ctools/service_container_test_ctools.info:7]: hidden = TRUE

is missing here.

**Comment by drupol** [comment]
It looks like as soons as there are upper cases in the plugin type name or in the plugin name, it fails.
Anyway, I'll take some rest, it's driving me crazy !

**Comment by drupol** [comment]
Question for later: How to deal with plugins that doesn't provide class ? Like [ctools.cache.export_ui](http://cgit.drupalcode.org/ctools/tree/plugins/cache/export_ui.inc)

**Comment by fabianx-ai** [comment]
Closing in favor of #35

**Pull request closed without merging**